File: src/main/python/RefactoredOutput.py
The user wants to add a feedback mechanism after they've used the analysis tools. Given the file structure, `pages/sales_forecasting.py` is a strong candidate for an analysis tool, and `db.py` and `auth.py` can be used to store and associate feedback with users.

Here's a plan to implement this:

1.  **Modify `db.py`**: Add functions to create a `feedback` table and to insert feedback.
2.  **Integrate into an analysis page (e.g., `pages/sales_forecasting.py`)**:
    *   Import necessary `auth` and `db` functions.
    *   Add a Streamlit form for feedback (rating and comments) at the end of the analysis section.
    *   Ensure feedback is only shown to logged-in users.

Let's proceed with the changes.

---

**Proposed Changes:**

**1. Modify `db.py` to handle feedback storage**

This will add functions to create a `feedback` table if it doesn't exist and to store new feedback entries.

```python
# db.py
import sqlite3
import hashlib

DATABASE = 'users.db'

def get_db_connection():
    conn = sqlite3.connect(DATABASE)
    conn.row_factory = sqlite3.Row  # This allows accessing columns by name
    return conn

def create_users_table():
    conn = get_db_connection()
    c = conn.cursor()
    c.execute('''
        CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            username TEXT NOT NULL UNIQUE,
            password TEXT NOT NULL
        )
    ''')
    conn.commit()
    conn.close()

# New function: Create feedback table
def create_feedback_table():
    conn = get_db_connection()
    c = conn.cursor()
    c.execute('''
        CREATE TABLE IF NOT EXISTS feedback (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            user_id INTEGER NOT NULL,
            rating INTEGER,
            comments TEXT,
            timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
            FOREIGN KEY (user_id) REFERENCES users(id)
        )
    ''')
    conn.commit()
    conn.close()

def add_user(username, password):
    conn = get_db_connection()
    c = conn.cursor()
    hashed_password = hashlib.sha256(password.encode()).hexdigest()
    try:
        c.execute("INSERT INTO users (username, password) VALUES (?, ?)", (username, hashed_password))
        conn.commit()
        return True
    except sqlite3.IntegrityError:
        return False # Username already exists
    finally:
        conn.close()

def verify_user(username, password):
    conn = get_db_connection()
    c = conn.cursor()
    hashed_password = hashlib.sha256(password.encode()).hexdigest()
    c.execute("SELECT * FROM users WHERE username = ? AND password = ?", (username, hashed_password))
    user = c.fetchone()
    conn.close()
    return user

def get_user_id(username):
    conn = get_db_connection()
    c = conn.cursor()
    c.execute("SELECT id FROM users WHERE username = ?", (username,))
    user = c.fetchone()
    conn.close()
    return user['id'] if user else None

# New function: Add feedback
def add_feedback(user_id, rating, comments):
    conn = get_db_connection()
    c = conn.cursor()
    c.execute("INSERT INTO feedback (user_id, rating, comments) VALUES (?, ?, ?)",
              (user_id, rating, comments))
    conn.commit()
    conn.close()

# Ensure tables are created when db.py is imported or application starts
create_users_table()
create_feedback_table() # Call this to ensure the table exists
```

**2. Modify `pages/sales_forecasting.py` to add the feedback form**

We'll add a feedback section at the bottom of the `pages/sales_forecasting.py` page, assuming this is where users would complete their analysis. We'll check if the user is logged in before showing the form.

```python
# pages/sales_forecasting.py
import streamlit as st
import pandas as pd
import numpy as np
import plotly.express as px
from datetime import datetime, timedelta
import auth
import db # Import the db module

# --- Helper functions (assuming they are already in utils.py or defined here) ---
def generate_sample_sales_data(days=90):
    dates = [datetime.now() - timedelta(days=i) for i in range(days-1, -1, -1)]
    sales = np.random.randint(50, 200, days) + np.sin(np.linspace(0, 3*np.pi, days)) * 50
    data = pd.DataFrame({'Date': dates, 'Sales': sales.round(0)})
    return data

def forecast_sales(df, periods=30):
    # Simple moving average for demonstration
    df['SMA_7'] = df['Sales'].rolling(window=7).mean().shift(1)
    df['SMA_30'] = df['Sales'].rolling(window=30).mean().shift(1)

    last_date = df['Date'].max()
    forecast_dates = [last_date + timedelta(days=i) for i in range(1, periods + 1)]

    # Use the last known SMA_30 value for a simple forecast
    last_sma_30 = df['SMA_30'].iloc[-1]
    forecast_sales_values = [last_sma_30 + np.random.randint(-10, 10) for _ in range(periods)] # Add some noise

    forecast_df = pd.DataFrame({'Date': forecast_dates, 'Sales': forecast_sales_values})
    return forecast_df

# --- Streamlit Page Content ---
st.set_page_config(page_title="Sales Forecasting", layout="wide")

st.title("📊 Sales Forecasting Tool")

if auth.check_login():
    st.write(f"Welcome, {st.session_state['username']}!")

    st.subheader("Upload Your Sales Data")
    uploaded_file = st.file_uploader("Choose a CSV file", type="csv")

    df = None
    if uploaded_file is not None:
        try:
            df = pd.read_csv(uploaded_file)
            st.success("File uploaded successfully!")
            st.write("Preview of your data:")
            st.dataframe(df.head())

            # Basic data cleaning/preparation (assuming 'Date' and 'Sales' columns)
            if 'Date' in df.columns and 'Sales' in df.columns:
                df['Date'] = pd.to_datetime(df['Date'])
                df = df.sort_values('Date').set_index('Date')
                df['Sales'] = pd.to_numeric(df['Sales'])
                st.write("Data processed and ready for forecasting.")
            else:
                st.error("Please ensure your CSV has 'Date' and 'Sales' columns.")
                df = None
        except Exception as e:
            st.error(f"Error reading file: {e}")
            df = None
    else:
        st.info("No file uploaded. Generating sample sales data for demonstration.")
        df = generate_sample_sales_data()
        df = df.set_index('Date')


    if df is not None and not df.empty:
        st.subheader("Historical Sales Data")
        fig_hist = px.line(df, x=df.index, y='Sales', title="Historical Sales Trend")
        st.plotly_chart(fig_hist, use_container_width=True)

        st.subheader("Forecast Future Sales")
        periods = st.slider("Number of days to forecast", 7, 365, 30)

        if st.button("Generate Forecast"):
            forecast_df = forecast_sales(df.reset_index(), periods)

            st.write("Forecasted Sales:")
            st.dataframe(forecast_df)

            combined_df = pd.concat([df.reset_index(), forecast_df])
            fig_forecast = px.line(combined_df, x='Date', y='Sales',
                                   title="Historical and Forecasted Sales")
            fig_forecast.add_vline(x=df.index.max(), line_dash="dash", line_color="red",
                                    annotation_text="Forecast Start", annotation_position="top left")
            st.plotly_chart(fig_forecast, use_container_width=True)

            st.success("Sales forecast generated successfully!")
            st.session_state['analysis_complete'] = True # Set a flag that analysis is done
    else:
        st.warning("Please upload a valid dataset or use the sample data to proceed with forecasting.")

    # --- New Feedback Section ---
    st.markdown("---") # Separator
    st.header("We'd Love Your Feedback! 💡")
    st.write("Help us improve by sharing your thoughts on this analysis tool. Your input is highly valued!")

    # Check if analysis was completed before prompting for feedback for the first time
    # This uses a session state variable to ensure feedback is prompted after a successful analysis
    if st.session_state.get('analysis_complete', False):
        user_info = auth.get_current_user_info() # Assuming auth.py has a function to get user details
        if user_info and 'username' in user_info:
            user_id = db.get_user_id(user_info['username'])

            if user_id:
                # Use a form to group inputs and clear on submit
                with st.form("feedback_form", clear_on_submit=True):
                    st.subheader("How would you rate your experience with this tool?")
                    rating = st.slider("Rating", 1, 5, 3, help="1 = Very Poor, 5 = Excellent")
                    comments = st.text_area("Your Comments", height=150,
                                            help="Tell us what you liked, disliked, or what could be improved.")

                    submitted = st.form_submit_button("Submit Feedback")

                    if submitted:
                        if comments.strip(): # Ensure comments are not just whitespace
                            try:
                                db.add_feedback(user_id, rating, comments)
                                st.success("Thank you for your feedback! It helps us a lot. 🙏")
                                # Optionally, prevent submitting multiple feedbacks in the same session for the same analysis
                                st.session_state['feedback_submitted_for_session'] = True
                            except Exception as e:
                                st.error(f"Failed to submit feedback: {e}")
                        else:
                            st.warning("Please provide some comments before submitting your feedback.")
            else:
                st.error("Could not retrieve user ID for feedback submission. Please try logging in again.")
        else:
            st.error("User information not available for feedback submission.")
    else:
        st.info("Complete an analysis first to provide feedback.")


else:
    st.warning("Please log in to access the Sales Forecasting tool and provide feedback.")
```

**3. Modify `auth.py` (if `get_current_user_info` doesn't exist)**

The `pages/sales_forecasting.py` code assumes `auth.get_current_user_info()` exists. If it doesn't, you'll need to add it to `auth.py` to retrieve the current logged-in user's details, especially their username, which is then used to get their `user_id` from the `db`.

```python
# auth.py (Ensure this function exists or adapt to your auth system)
import streamlit as st
import db

def check_login():
    if 'logged_in' not in st.session_state or not st.session_state['logged_in']:
        # st.warning("Please log in to access this page.")
        return False
    return True

def login(username, password):
    user = db.verify_user(username, password)
    if user:
        st.session_state['logged_in'] = True
        st.session_state['username'] = username
        return True
    return False

def logout():
    st.session_state['logged_in'] = False
    del st.session_state['username']
    if 'analysis_complete' in st.session_state:
        del st.session_state['analysis_complete']
    if 'feedback_submitted_for_session' in st.session_state:
        del st.session_state['feedback_submitted_for_session']
    st.success("You have been logged out.")
    st.rerun()

# New function: Get current user info
def get_current_user_info():
    if check_login():
        return {'username': st.session_state['username']}
    return None
```

---

**Summary of Changes and How to Use:**

1.  **`db.py`**:
    *   A new function `create_feedback_table()` is added to define the `feedback` table. This table stores `user_id`, `rating`, `comments`, and `timestamp`.
    *   A new function `add_feedback(user_id, rating, comments)` is added to insert feedback into the database.
    *   `create_feedback_table()` is called directly in `db.py` to ensure the table is created when the module is imported or the app starts.

2.  **`pages/sales_forecasting.py`**:
    *   Imports `db` and `auth`.
    *   A `st.session_state['analysis_complete']` flag is set to `True` after a forecast is successfully generated. This ensures the feedback form only appears once the user has actually used the tool.
    *   A new section for "We'd Love Your Feedback!" is added at the bottom of the page.
    *   This section uses `st.form` for collecting feedback (a slider for rating, a text area for comments).
    *   It retrieves the `user_id` of the currently logged-in user using `auth.get_current_user_info()` and `db.get_user_id()`.
    *   Upon submission, it calls `db.add_feedback()` to store the data and provides a success message.
    *   It checks for `auth.check_login()` to ensure only logged-in users can see and submit feedback.
    *   A `st.session_state['feedback_submitted_for_session']` flag is added (optional) to prevent accidental multiple submissions in a single session for the *same analysis run*.

3.  **`auth.py`**:
    *   A `get_current_user_info()` function is added to provide the current username, which is then used to fetch the user ID from the database.

**To run this:**

1.  Ensure your Streamlit application is running.
2.  Log in with a registered user (you might need to register via `pages/register.py` first).
3.  Navigate to the "Sales Forecasting" page.
4.  Upload a CSV or use the sample data and generate a forecast.
5.  After the forecast is displayed, scroll down to see the "We'd Love Your Feedback!" section.
6.  Provide a rating and comments, then click "Submit Feedback."
7.  The feedback will be stored in your `users.db` file. You can inspect `users.db` with an SQLite browser to verify the `feedback` table and its contents.

File: src/main/java/RefactoredOutput.java
To add a feature that asks the user about the effectiveness of the tool after they analyze their sales file, we'll implement the following steps:

1.  **Modify `db.py`**: Add functions to create a new `feedback` table and to insert feedback data into it.
2.  **Modify `auth.py`**: Ensure we can retrieve the `user_id` of the currently logged-in user.
3.  **Modify `pages/sales_forecasting.py`**: This is where the sales analysis likely happens. We will add the feedback form after the analysis results are displayed.

---

### Step 1: Update `db.py` to handle feedback storage

We need to add a new table to `users.db` to store the feedback.

**File: `db.py`**

```python
import sqlite3
import hashlib # Assuming this is used for password hashing in auth.py if not already imported

DATABASE_FILE = 'users.db'

def get_db_connection():
    conn = sqlite3.connect(DATABASE_FILE)
    conn.row_factory = sqlite3.Row  # This allows accessing columns by name
    return conn

# Existing functions in db.py (e.g., create_users_table, add_user, check_user, etc.)
# ...

# --- NEW FUNCTIONS FOR FEEDBACK ---

def create_feedback_table():
    """Creates the feedback table if it doesn't already exist."""
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS feedback (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            user_id INTEGER,
            feature_name TEXT NOT NULL,
            effectiveness_score INTEGER NOT NULL,
            comments TEXT,
            timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
            FOREIGN KEY (user_id) REFERENCES users (id) ON DELETE SET NULL
        )
    ''')
    conn.commit()
    conn.close()

def insert_feedback(user_id, feature_name, effectiveness_score, comments):
    """Inserts a new feedback entry into the database."""
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute(
        'INSERT INTO feedback (user_id, feature_name, effectiveness_score, comments) VALUES (?, ?, ?, ?)',
        (user_id, feature_name, effectiveness_score, comments)
    )
    conn.commit()
    conn.close()
    
# Call create_feedback_table() when the database module is imported, 
# ensuring the table exists. This is a common pattern for Streamlit apps.
create_feedback_table() 
```

**Explanation of changes in `db.py`:**
*   `create_feedback_table()`: This function defines and creates a new table named `feedback`.
    *   `id`: Primary key for unique feedback entries.
    *   `user_id`: Foreign key referencing the `users` table, allowing us to link feedback to a specific user. `ON DELETE SET NULL` ensures feedback remains even if a user account is deleted.
    *   `feature_name`: To specify which feature the feedback is for (e.g., "Sales Forecasting").
    *   `effectiveness_score`: An integer representing the rating (e.g., 1-5).
    *   `comments`: An optional text field for detailed feedback.
    *   `timestamp`: Automatically records when the feedback was submitted.
*   `insert_feedback()`: A utility function to easily add new feedback records.
*   `create_feedback_table()` is called directly when `db.py` is imported. This ensures the table is created early in the application lifecycle.

---

### Step 2: Update `auth.py` to get current user ID

We need the `user_id` to link the feedback to a specific user. Assuming your `auth.py` handles user sessions with `st.session_state`, we'll add a helper function.

**File: `auth.py`**

```python
import streamlit as st
import hashlib
from db import get_db_connection # Assuming get_db_connection is in db.py

# ... (Existing login, register, logout functions) ...

def get_current_username():
    """Returns the username of the currently logged-in user from session state."""
    return st.session_state.get('username')

def get_current_user_id():
    """
    Returns the user ID of the currently logged-in user from session state.
    Assumes 'user_id' is stored in st.session_state upon successful login.
    If not, it fetches it from the database using the username.
    """
    if 'user_id' in st.session_state and st.session_state['user_id'] is not None:
        return st.session_state['user_id']
    else:
        # Fallback: if user_id is not directly in session, try to get it from username
        username = get_current_username()
        if username:
            conn = get_db_connection()
            cursor = conn.cursor()
            cursor.execute('SELECT id FROM users WHERE username = ?', (username,))
            user = cursor.fetchone()
            conn.close()
            if user:
                # Store it in session state for future use
                st.session_state['user_id'] = user['id']
                return user['id']
    return None

# Make sure your login function (e.g., in auth.py) populates st.session_state['user_id']
# Example (if your login function looks similar):
# def login_user(username, password):
#     conn = get_db_connection()
#     cursor = conn.cursor()
#     cursor.execute('SELECT id, username, password FROM users WHERE username = ?', (username,))
#     user = cursor.fetchone()
#     conn.close()
#     if user and check_password_hash(user['password'], password): # Assuming check_password_hash exists
#         st.session_state['logged_in'] = True
#         st.session_state['username'] = username
#         st.session_state['user_id'] = user['id'] # IMPORTANT: Store user_id here
#         return True
#     return False

```

**Explanation of changes in `auth.py`:**
*   `get_current_user_id()`: This function retrieves the `user_id` from Streamlit's `session_state`. It includes a fallback to fetch it from the database if it's not directly present in `session_state`, though it's best practice to store it there during login.
*   **Important:** Ensure your existing `login_user` function (likely in `auth.py`) stores the `user_id` in `st.session_state` when a user successfully logs in.

---

### Step 3: Add feedback form to `pages/sales_forecasting.py`

This is where the user interacts with the sales analysis. We'll present the feedback form *after* they've performed an analysis.

**File: `pages/sales_forecasting.py`**

```python
import streamlit as st
import pandas as pd
# Assuming db and auth are in the root directory or correctly imported via path
from db import insert_feedback 
from auth import get_current_user_id, get_current_username # Import necessary auth functions

st.set_page_config(page_title="Sales Forecasting", page_icon="📈")

st.title("📈 Sales Forecasting and Analysis")

# --- Existing logic for sales file analysis and forecasting ---
# This part of your code would involve uploading files, processing them,
# displaying charts, tables, and forecasts.
# We need a way to determine when this analysis is "complete".
# For demonstration, let's use a session state variable.

if 'sales_analysis_completed' not in st.session_state:
    st.session_state['sales_analysis_completed'] = False

# Simulate sales analysis completion
if st.button("Perform Sales Analysis (Simulated)"):
    # In a real scenario, this would be triggered after data processing
    # and model execution.
    st.write("Performing complex sales analysis...")
    st.write("Displaying forecast results...")
    # Example dummy results:
    st.success("Sales analysis complete! Here are your forecasts.")
    st.line_chart(pd.DataFrame({'Actual': [10, 12, 15, 13, 16], 'Forecast': [11, 13, 14, 14, 15]}))
    st.session_state['sales_analysis_completed'] = True
    # Reset feedback submission state for a new analysis session
    st.session_state['feedback_submitted_for_this_analysis'] = False

# Only show the feedback form if the analysis has been completed
if st.session_state.get('sales_analysis_completed'):
    st.markdown("---") # Separator for clarity
    st.subheader("We value your feedback!")
    st.write("How effective did you find our sales forecasting tool for your needs?")

    # Define effectiveness options and their corresponding scores
    feedback_options = {
        "1 - Very Ineffective": 1,
        "2 - Ineffective": 2,
        "3 - Neutral": 3,
        "4 - Effective": 4,
        "5 - Very Effective": 5
    }
    
    # Check if feedback has already been submitted for this analysis session
    if not st.session_state.get('feedback_submitted_for_this_analysis', False):
        selected_effectiveness_text = st.radio(
            "Select your rating:",
            list(feedback_options.keys()),
            index=2, # Default to "3 - Neutral"
            key="effectiveness_rating_radio" # Unique key for this widget
        )
        effectiveness_score = feedback_options[selected_effectiveness_text]

        comments = st.text_area("Optional: Any specific comments or suggestions?", key="feedback_comments_area")

        if st.button("Submit Feedback", key="submit_feedback_button"):
            user_id = get_current_user_id() # Get the logged-in user's ID
            username = get_current_username() # For logging/debug if needed

            if user_id is None:
                st.warning("Please log in to submit feedback.")
            else:
                try:
                    insert_feedback(
                        user_id=user_id,
                        feature_name="Sales Forecasting", # Specific to this page/feature
                        effectiveness_score=effectiveness_score,
                        comments=comments
                    )
                    st.success("Thank you for your valuable feedback! We appreciate it.")
                    st.session_state['feedback_submitted_for_this_analysis'] = True
                    # Optionally clear the form elements visually
                    st.session_state["effectiveness_rating_radio"] = "3 - Neutral" # Reset radio
                    st.session_state["feedback_comments_area"] = "" # Clear text area
                except Exception as e:
                    st.error(f"Failed to submit feedback. Please try again. Error: {e}")
    else:
        st.info("You've already submitted feedback for this analysis session. Thank you!")

# --- Rest of your sales_forecasting.py content ---
# ... (e.g., sidebar, other analysis components) ...

```

**Explanation of changes in `pages/sales_forecasting.py`:**
*   **Imports**: Import `insert_feedback` from `db` and `get_current_user_id`, `get_current_username` from `auth`.
*   **`st.session_state['sales_analysis_completed']`**: A flag to control when the feedback form appears. It should be set to `True` *after* the sales analysis results are presented to the user. I've added a dummy button to simulate this.
*   **Feedback Form**:
    *   `st.subheader` and `st.write` introduce the feedback section.
    *   `st.radio` provides a Likert scale (1-5) for effectiveness.
    *   `st.text_area` allows for optional detailed comments.
    *   `st.button("Submit Feedback")` triggers the submission process.
*   **Submission Logic**:
    *   When the button is clicked, `get_current_user_id()` fetches the ID of the logged-in user.
    *   `insert_feedback()` is called to save the data to the `feedback` table.
    *   Success/error messages (`st.success`, `st.error`) provide user feedback.
    *   `st.session_state['feedback_submitted_for_this_analysis']` is used to prevent multiple submissions for the same analysis results in a single session.

---

### To run and test:

1.  **Ensure all files are saved** with the changes.
2.  **Make sure your `users.db` exists** and has a `users` table (created by your `auth.py`/`db.py` on first run usually).
3.  **Start your Streamlit app**: `streamlit run Home.py` (or `streamlit run pages/login.py` if that's your entry for pages).
4.  **Log in** with a registered user.
5.  **Navigate to the "Sales Forecasting" page**.
6.  **Click the "Perform Sales Analysis (Simulated)" button** (or whichever action in your actual code triggers the analysis completion).
7.  **The feedback form should appear**. Select a rating, add comments, and click "Submit Feedback".
8.  **Verify**: You can use a SQLite browser (like DB Browser for SQLite) to open `users.db` and check if the `feedback` table has been created and contains your submitted data.